transport/transport.go

```diff
@@ -120,43 +125,29 @@ func NewTransport(conn *websocket.Conn) Transport {
 	}
 }
 
+// Close attempts to send a "going away" message over the websocket connection.
+// This will cause a Write over the websocket transport, which can cause a
+// panic. We rescue potential panics and consider the connection closed,
+// returning nil, because the connection _will_ be closed. Hay!
+func (t *WebSocketTransport) Close() error {
+	t.mutex.Lock()
+	defer func() {
```

Comment:
Why aren't we locking the mutex with every `WriteMessage`?

Reply:
Basically the whole transport needs reworking. This is some of the oldest code in 2.0.

Reply:
(There's tight coupling that shouldn't exist that makes things "generally work."